### PR TITLE
Add module project_task_link

### PR DESCRIPTION
**Pull request by ddufresne**
https://www.numigi.com/web#id=17193&action=298&model=project.task&view_type=form&menu_id=200

### CODE REVIEW

File: project_task_link/__init__.py [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# © 2019 Numigi (tm) and all its contributors (https://bit.ly/numigiens)
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+from . import models
```
Comment (ddufresne) [inline project_task_link/__init__.py]: Codacy Issue found: ['.models' imported but unused](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=4478774)

File: project_task_link/models/common.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+import re
+from lxml import etree
+
+
+DEFAULT_TASK_REF_REGEX = r'[Tt][Aa]\#?\d+'
+DEFAULT_TASK_REF_FORMAT = "TA#{task_id}"
+
+
+def _get_task_ref_regex(env: 'Environment') -> str:
```
Comment (ddufresne) [inline project_task_link/models/common.py]: Codacy Issue found: [invalid syntax](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=4478774)

File: project_task_link/static/src/js/html_field.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
+/*
+    © 2019 Numigi (tm) and all its contributors (https://bit.ly/numigiens)
+    License LGPL-3.0 or later (http://www.gnu.org/licenses/LGPL.html).
+*/
+odoo.define("project_task_link.html_field", function(require) {
+"use strict";
+
+var HtmlField = require('web_editor.backend').FieldTextHtmlSimple;
```
Comment (ddufresne) [inline project_task_link/static/src/js/html_field.js]: Codacy Issue found: [Strings must use doublequote. (quotes)](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=4478774)

File: project_task_link/models/__init__.py [hunk 4]
```diff
@@ -0,0 +1,4 @@
+# © 2019 Numigi (tm) and all its contributors (https://bit.ly/numigiens)
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+from . import project_task
```
Comment (ddufresne) [inline project_task_link/models/__init__.py]: Codacy Issue found: ['.project_task' imported but unused](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=4478774)

File: project_task_link/static/src/js/html_field.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+            params: {
+                context: odoo.session_info.user_context,
+            },
+        });
+        this.trigger_up("do_action", {action});
+    },
+    _renderReadonly(){
+        this._super.apply(this, arguments);
+        var self = this;
+        var taskLinks = this.$('a').filter(function() {
```
Comment (ddufresne) [inline project_task_link/static/src/js/html_field.js]: Codacy Issue found: [Strings must use doublequote. (quotes)](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=4478774)

File: project_task_link/static/src/js/html_field.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
+/*
+    © 2019 Numigi (tm) and all its contributors (https://bit.ly/numigiens)
+    License LGPL-3.0 or later (http://www.gnu.org/licenses/LGPL.html).
+*/
+odoo.define("project_task_link.html_field", function(require) {
```
Comment (ddufresne) [inline project_task_link/static/src/js/html_field.js]: Codacy Issue found: ['odoo' is not defined. (no-undef)](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=4478774)

File: project_task_link/models/common.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
+# © 2019 Numigi (tm) and all its contributors (https://bit.ly/numigiens)
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+import re
+from lxml import etree
```
Comment (ddufresne) [inline project_task_link/models/common.py]: Codacy Issue found: [Using etree to parse untrusted XML data is known to be vulnerable to XML attacks. Replace etree with the equivalent defusedxml package.](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=4478774)

File: project_task_link/static/src/js/html_field.js [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    /**
+     * Open the form view of a task given the task ID.
+     */
+    async _openTaskFormView(taskId){
+        var action = await this._rpc({
+            model: "project.task",
+            method: "get_formview_action",
+            args: [[taskId]],
+            params: {
+                context: odoo.session_info.user_context,
```
Comment (ddufresne) [inline project_task_link/static/src/js/html_field.js]: Codacy Issue found: ['odoo' is not defined. (no-undef)](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=4478774)

File: project_task_link/README.rst [hunk 9]
```diff
@@ -0,0 +1,67 @@
+Project Task Link
+=================
+This module allows to insert dynamic links in the descriptions of tasks.
```
Comment (foutoucour) [inline project_task_link/README.rst]:
```suggestion
This module allows to insert dynamic links in the description of tasks.
```

File: project_task_link/tests/test_project_task.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+    @data(
+        "Refer to <a>{task_ref}<a> for more information.",
+        "Refer to <a><button>{task_ref}</button><a> for more information.",
+    )
+    def test_if_task_ref_inside_link_node__no_link_node_added(self, description):
+        self.task.write({'description': description.format(task_ref=self.task_ref)})
+        assert self.task_link not in self.task.description
+
+    @data(
+        "TA{}",
```
Comment (foutoucour) [inline project_task_link/tests/test_project_task.py:68]:
```suggestion
        "TA{}",
        "TA#{}",
```

File: project_task_link/models/common.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+
+DEFAULT_TASK_REF_REGEX = r'[Tt][Aa]\#?\d+'
```
Comment (foutoucour) [inline project_task_link/models/common.py]: Je suis partagé sur cette implémentation. 
Je suis partagé car nous allons avoir besoin de la même logique pour Github PR à un moment,
et je trouve le tout compliqué.

Je comprends que ça permet l'utilisation d'un ir.config.parameter pour rendre custom la regex, mais est-ce la bonne manière de répondre au besoin, je ne suis pas sûr.

Je te propose un chat de 15-30 mins demain sur le sujet, ok ?

**Review by foutoucour — CHANGES_REQUESTED** [review]

File: project_task_reference/models/project_task.py [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+import re
+from odoo import api, models, _
+from odoo.exceptions import MissingError
+from ..reference import TaskReference
+
+DEFAULT_TASK_REF_REGEX = r'[tT][aA]#?(?P<id>\d+)'
+DEFAULT_TASK_REF_FORMAT = "TA#{id}"
+
+
+def _get_task_ref_regex(env: 'Environment') -> str:
```
Comment (ddufresne) [inline project_task_reference/models/project_task.py]: Codacy Issue found: [invalid syntax](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=4478774)

File: project_task_reference/tests/test_project_task.py [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+        text = "Some prefix text {} some suffix text".format(expected_string)
+        reference = self.env['project.task']._extract_references_from_text(text)[0]
+        assert reference.string == expected_string
+
+    def test_reference_normalized_string(self):
+        text = "Some prefix text {} some suffix text".format("ta#{}".format(self.task_1.id))
+        reference = self.env['project.task']._extract_references_from_text(text)[0]
+        assert reference.normalized_string == "TA#{}".format(self.task_1.id)
+
+    def _make_reference_from_task_id(self, task_id):
```
Comment (ddufresne) [inline project_task_reference/tests/test_project_task.py]: Codacy Issue found: [Method could be a function](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=4478774)

File: project_task_reference/__init__.py [hunk 14]
```diff
@@ -0,0 +1,4 @@
+# © 2019 Numigi (tm) and all its contributors (https://bit.ly/numigiens)
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+from . import models
```
Comment (ddufresne) [inline project_task_reference/__init__.py]: Codacy Issue found: ['.models' imported but unused](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=4478774)

File: project_task_link/models/common.py [hunk 15]
```diff
@@ -0,0 +1,51 @@
+# © 2019 Numigi (tm) and all its contributors (https://bit.ly/numigiens)
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+from lxml import etree
+from odoo.addons.project_task_reference.reference import TaskReference
+from typing import List
+
+
+def _find_task_references(env: 'Environment', text: str) -> List[TaskReference]:
```
Comment (ddufresne) [inline project_task_link/models/common.py]: Codacy Issue found: [invalid syntax](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=4478774)

File: project_task_reference/reference.py [hunk 16]
```diff
@@ -0,0 +1,33 @@
+# © 2019 Numigi (tm) and all its contributors (https://bit.ly/numigiens)
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+from typing import Mapping
+
+
+class TaskReference:
+
+    def __init__(self, values: Mapping[str, str], string: str, format_: str):
```
Comment (ddufresne) [inline project_task_reference/reference.py]: Codacy Issue found: [invalid syntax](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=4478774)

File: project_task_reference/models/project_task.py [hunk 17]
```diff
@@ -0,0 +1,74 @@
+# © 2019 Numigi (tm) and all its contributors (https://bit.ly/numigiens)
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+import re
+from odoo import api, models, _
```
Comment (ddufresne) [inline project_task_reference/models/project_task.py]: Codacy Issue found: [Unused api imported from odoo](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=4478774)

File: project_task_reference/__init__.py (hunk 14 above)
Comment (ddufresne) [inline project_task_reference/__init__.py]: Codacy Issue found: ['.models' imported but unused](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=4478774)

File: project_task_reference/reference.py (hunk 16 above)
Comment (ddufresne) [inline project_task_reference/reference.py]: Codacy Issue found: [invalid syntax](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=4478774)

File: project_task_reference/models/project_task.py (hunk 12 above)
Comment (ddufresne) [inline project_task_reference/models/project_task.py]: Codacy Issue found: [invalid syntax](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=4478774)

File: project_task_reference/tests/test_project_task.py (hunk 13 above)
Comment (ddufresne) [inline project_task_reference/tests/test_project_task.py]: Codacy Issue found: [Method could be a function](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=4478774)

File: project_task_link/models/common.py (hunk 15 above)
Comment (ddufresne) [inline project_task_link/models/common.py]: Codacy Issue found: [invalid syntax](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=4478774)

File: project_task_reference/models/project_task.py (hunk 17 above)
Comment (ddufresne) [inline project_task_reference/models/project_task.py]: Codacy Issue found: [Unused api imported from odoo](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=4478774)

**Comment by foutoucour** [comment]
@ddufresne ça fail :D

File: project_task_link/models/common.py (hunk 15 above)
Comment (ddufresne) [inline project_task_link/models/common.py]: Codacy Issue found: [invalid syntax](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=4478774)

File: project_task_reference/models/project_task.py (hunk 12 above)
Comment (ddufresne) [inline project_task_reference/models/project_task.py]: Codacy Issue found: [invalid syntax](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=4478774)

File: project_task_reference/reference.py (hunk 16 above)
Comment (ddufresne) [inline project_task_reference/reference.py]: Codacy Issue found: [invalid syntax](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=4478774)

File: project_task_reference/tests/test_project_task.py (hunk 13 above)
Comment (ddufresne) [inline project_task_reference/tests/test_project_task.py]: Codacy Issue found: [Method could be a function](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=4478774)

File: project_task_reference/__init__.py (hunk 14 above)
Comment (ddufresne) [inline project_task_reference/__init__.py]: Codacy Issue found: ['.models' imported but unused](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=4478774)

File: project_task_reference/models/project_task.py (hunk 17 above)
Comment (ddufresne) [inline project_task_reference/models/project_task.py]: Codacy Issue found: [Unused api imported from odoo](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=4478774)

**Review by foutoucour — APPROVED** [review]

**Pull request merged**